Add matrix math module

Mridul97 opened the pull request:
Issue #324 
This module takes pixels of the input picture and the 9 values of 3x3 kernel as arguments and returns the pixels of the picture after applying convolution.

For an array [1/9, 1/9, 1/9, 1/9, 1/9, 1/9, 1/9, 1/9, 1/9] and pixels of picture on the left as input, the module returns pixels of the picture on the right.

![ss_9](https://user-images.githubusercontent.com/21158457/46255890-ae30d180-c4c0-11e8-9d27-66144f0f5c00.jpg)

tech4GT commented:
@Mridul97 This would support any kernel right? Could you please attach a output screenshot fot his 
```js
[
 [1,2,1],
 [0,0,0],
 [-1,-2,-1]
]
```

tech4GT commented:
Okay so Mridul couple of things here
- Please add a info.json file for the module as well, right now tit is not complete.
- Could you rename this module to convolution, that seems more apt

Mridul97 commented:
Hi, I have added the info.json file and also renamed the module to Convolution. Please have a look!

Yes, this will support any kernel of size 3x3. Output screenshot for the given kernel :-

![ss_10](https://user-images.githubusercontent.com/21158457/46291014-0217e500-c5ab-11e8-804e-30a22ac4af50.jpg)

tech4GT commented:
@Mridul97 Oh! Does this module apply convolution to all channels separately? Do you mind posting a screenshot with using this on a grayscale image, that should give horizontal edges in the image. Thanks!

Mridul97 commented:
Yes, this module applies convolution to all channels separately.

I tried to use it on a grayscale image using the given matrix as kernel. This is the output I am getting :-

![ss_12](https://user-images.githubusercontent.com/21158457/46332700-2cb17e80-c63b-11e8-9b09-fcce0b5a354f.jpg)

I think this is not the desired output:sweat_smile:

tech4GT commented:
Hmm let me see!

tech4GT commented:
@Mridul97 Please also add this module to the Modules.js file so that it can be visible to sequencer, how did you manage to run this without exporting it?

tech4GT commented on src/modules/Convolution/info.json:
```diff
@@ -0,0 +1,14 @@
+            "type": "an array of 9 number values",
+            "desc": "values of the kernel which is used as mask for Convolution",
+        },
+        "pixels": {
```
Shouldn't this be available to the module from sequencer itself?

Mridul97 replied:
Yes, the pixels but values of the kernel will be taken from user. Sorry for that, will make the required changes!

tech4GT commented on src/modules/Convolution/info.json:
```diff
@@ -0,0 +1,14 @@
+    "description": "Image Convolution using a given 3x3 kernel matrix",
+    "inputs": {
+        "kernelValues": {
+            "type": "an array of 9 number values",
```
I don't think we can take input type array, maybe we can do a string with nine space separated values.

Mridul97 replied:
Okay I will do that! Thanks!

tech4GT commented on src/modules/Convolution/Module.js:
```diff
@@ -0,0 +1,63 @@
+		return arr;
+	}
+
+	function getNeighbouringPixelPositions(pixelPosition) {
```
I think we already have an API in place for this.

Mridul97 replied:
Could you please give a link to the API? Thanks!

tech4GT commented on src/modules/Convolution/info.json:
```diff
@@ -0,0 +1,11 @@
+    "description": "Image Convolution using a given 3x3 kernel matrix",
+    "inputs": {
+        "kernelValues": {
+            "type": "a string of space separated numbers",
```
One small change, make this String, and add this part in the description.
```js
"type": "String",
"desc": "nine space separated numbers representing the kernel values in left to right and top to bottom format."
```

Mridul97 replied:
Yes, we can have a constant factor as an input. It will be a different input field right? I will make the requested changes! Thanks!

Mridul97 commented:
I have made the changes. Please have a look!

tech4GT commented:
Mridul, This looks great!! Nice work!! I'll just run this and figure out why sobel didn't work.  Then we can merge🎉

tech4GT commented on src/modules/Convolution/Convolution.js:
```diff
@@ -0,0 +1,67 @@
+module.exports = exports = function(pixels, constantFactor, kernelValues){
+	let kernel = kernelGenerator(constantFactor, kernelValues), oldpix = pixels;
```
Aha! @Mridul97 Don't you think this will just copy the pixels object by reference? Shouldn't we use lodash here?

tech4GT replied:
Aha! @Mridul97 Don't you think this will just copy the pixels object by reference? Shouldn't we use lodash here?

Mridul97 commented:
I have made the changes. Please have a look!

This is the screenshot of the output :

![ss_14](https://user-images.githubusercontent.com/21158457/46424644-56f85e80-c757-11e8-8a53-b758f6e69f66.jpg)

Just one thing, the default values right now are set to be constantFactor = 1 and kernelValues = '0 0 0 0 0 0 0 0 0', it displays a black picture as output. I think we should change the default value of kernelValues. What should be the default value of kernelValues?

tech4GT commented:
@Mridul97 I agree, let's make it all values 1, it'll apply a little blur to the image! Also if you don't mind can you also post the screenshot with running this on a greyscale image, I am confident that this will work but just to be extra sure! Thanks!!

Mridul97 commented:
If we keep all the kernel values as 1, this is the output we are getting :
![ss_15](https://user-images.githubusercontent.com/21158457/46426138-4813ab00-c75b-11e8-9b33-63b38afa1073.jpg)

The output for a grayscale image :
Input:

![ss_16](https://user-images.githubusercontent.com/21158457/46426246-9b85f900-c75b-11e8-9b84-91ae38d621a6.jpg)

Output:

![ss_17](https://user-images.githubusercontent.com/21158457/46426298-c40df300-c75b-11e8-8867-516720f26dd4.jpg)

tech4GT commented:
@Mridul97 good thing we tried all 1s, we didn't normalise the values. So before calling pixels.set you should do a %255 on the values so they don't exceed the 0-255 range and for the default value keep the constant factor 1/9. Thanks!!

jywarren commented:
OOHHHHHH wow!

Great work you two!!!

On Wed, Oct 3, 2018 at 1:14 PM Varun Gupta <notifications@github.com> wrote:

> @Mridul97 <https://github.com/Mridul97> good thing we tried all 1s, we
> didn't normalise the values. So before calling pixels.set you should do a
> %255 on the values so they don't exceed the 0-255 range and for the default
> value keep the constant factor 1/9. Thanks!!
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/image-sequencer/pull/358#issuecomment-426720132>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABfJ47n-uEkQHZA9jjiZDtQeD2thWp0ks5uhPBFgaJpZM4XA5Mw>
> .
>

Mridul97 commented:
I have made the required changes. Please have a look!

tech4GT commented:
@jywarren we can do that but right now we take input as 9 space separated numbers in left to right and top to bottom format. So that should make a great follow up!

tech4GT commented:
@Mridul97 Please rebase this and then let's merge!

Mridul97 commented:
Yes, it does have a space in the UI to input the matrix parameter as a string. We can make a drop-down with some common matrices as follow-up. It will make UI more user friendly!

Screenshot of final output :
![ss_18](https://user-images.githubusercontent.com/21158457/46488027-27148e00-c81f-11e8-9be8-3a31ef3a425f.jpg)

Mridul97 commented:
I tried rebasing the branch but I got the following error :

![ss_19](https://user-images.githubusercontent.com/21158457/46488177-97bbaa80-c81f-11e8-8b18-5b9e4af31bac.jpg)

Could you please help me with this?

tech4GT commented:
There is also one other way, so instead of rebasing you can use `git pull https://github.com/publiclab/image-sequencer master 
This should land you in a merge state where you can open the conflicting files in an editor and resolve them manually. Then you can commit with a message like "merged" then push it.

Mridul97 commented:
Done! Thanks!

jywarren commented:
Yay!

jywarren commented:
This is very cool.

jywarren commented:
I'm publishing to the live demo now! This and any other modules should show up :-)
